src/platform/user/profile/utilities/index.js

```diff
@@ -141,3 +141,78 @@ export function teardownProfileSession() {
   sessionStorage.removeItem('shouldRedirectExpiredSession');
   clearSentryLoginType();
 }
+
+export const getValidationMessageKey = (
+  suggestedAddresses,
+  validationKey,
+  addressValidationError,
+) => {
+  let validationErrorKey;
+  const singleSuggestion = suggestedAddresses.length === 1;
+  const containsBadUnitNumber =
+    suggestedAddresses.filter(
+      address =>
+        address.addressMetaData?.deliveryPointValidation ===
+        'STREET_NUMBER_VALIDATED_BUT_BAD_UNIT_NUMBER',
+    ).length > 0;
+
+  const containsMissingUnitNumber =
+    suggestedAddresses.filter(
+      address =>
+        address.addressMetaData?.deliveryPointValidation ===
+        'STREET_NUMBER_VALIDATED_BUT_MISSING_UNIT_NUMBER',
+    ).length > 0;
+
+  if (singleSuggestion && containsBadUnitNumber) {
+    validationErrorKey = validationKey
+      ? 'badUnitNumberOverride'
+      : 'badUnitNumber';
+  }
+
+  if (singleSuggestion && containsMissingUnitNumber) {
+    validationErrorKey = validationKey
+      ? 'missingUnitNumberOverride'
+      : 'missingUnitNumber';
+  }
+
+  if (
+    singleSuggestion &&
+    !addressValidationError &&
+    !containsMissingUnitNumber &&
+    !containsBadUnitNumber
+  ) {
+    validationErrorKey = validationKey
+      ? 'showSuggestionsOverride'
+      : 'showSuggestions';
+  }
+
+  if (!singleSuggestion && !addressValidationError) {
+    validationErrorKey = validationKey
+      ? 'showSuggestionsOverride'
+      : 'showSuggestions';
+  }
+
+  if (addressValidationError) {
```

Comment:
I'm a fan of exiting functions early so I'd move this up and return from the function if it's true. Then you can remove all the `!addressValidationError` checks in the other conditionals.